### PR TITLE
generalize the portable MF methods to fe_degree < n_q_points_1d cases

### PR DESCRIPTION
**Pull request by YiminJin**
In this PR I generalize the portable matrix-free methods to three cases:

1. spectral elements with supporting points identical to the quadrature points (using collocation method);
2. elements that can be transformed to spectral elements;
3. elements with tensor-product shape functions that cannot use collocation method (or ineffective with collocation method).

The implementations are mostly imitated from the CPU version codes. Cases 2 and 3 have been tested with step-64 with fe_degree = 3 and n_q_points_1d = 4/5/6. The efficiency is almost the same as the original codes.

Comparing to the CPU version codes, my implementation is toy-like and incomplete (a lot of cases unconsidered, such as the elements with truncated tensor-product shape functions or cannot be represented with tensor-products). Some of the codes are even not tested (i.e., `apply()` with Kokkos version < 4.0). However, I need to run the matrix-free methods on GPUs in my research and I hope to get some advice on my implementation. Specifically, I have several questions:

1. In the original `apply()` function, the temporary array is created with `Number t[n_q_points]`. I do not know if it is allocated in the register or in the global memory. So I add a scratch pad in `Portable::SharedData` so that I can be sure that the temporary arrays are allocated in the shared memory (in case 3 we need more than 1 temporary arrays to reuse the intermediate results so the temporary storage may be large). Yet, it turns out that the computing efficiency has not changed. So, is the scratch pad in shared memory useful? And, furthermore, would it be worthwhile to copy the shape data from the global memory into the shared memory at the beginning of the loop?
2. In the CPU codes, there are 3 versions of the `apply()` function, corresponding to the even-odd method, the symmetric case and the general case. As far as I can see, the symmetric case is faster than the general case because it only needs to read a half of the shape data. However, I do not know if it is worth being implemented for GPU, since the computing task is not reduced. As for the even-odd method, the multiplying operations are halved, but there are more adding operations. I do not know how to choose the most efficiency method in different cases.

Would you please take a look at my codes and questions? @kronbichler @Rombur

### CODE REVIEW

**Comment by kronbichler** [comment]
@YiminJin thank you for tackling this problem, this has been on the todo list for a long time. Let me start with answering the questions first:
> In the original apply() function, the temporary array is created with Number t[n_q_points]. I do not know if it is allocated in the register or in the global memory. So I add a scratch pad in Portable::SharedData so that I can be sure that the temporary arrays are allocated in the shared memory (in case 3 we need more than 1 temporary arrays to reuse the intermediate results so the temporary storage may be large). Yet, it turns out that the computing efficiency has not changed. So, is the scratch pad in shared memory useful? 

I think that eliminating the local `t` array is definitely the right way to go, so I fully support this change to more explicitly place the data in the right memory type. I must admit that I do not entirely understand the question: Are you saying that you measured the old code and your suggested change, and did not observe a difference in performance? I guess the important question to check is whether the performance reached some point of saturation before/after your change. For example, on a test I did on an A100 the code currently on dealii-master is very slow and far away from a saturation point, but this depends on the architecture, the compiler and specific flags.

> And, furthermore, would it be worthwhile to copy the shape data from the global memory into the shared memory at the beginning of the loop?

I think that would need to be measured, but I would think it could make sense indeed.

> In the CPU codes, there are 3 versions of the apply() function, corresponding to the even-odd method, the symmetric case and the general case. As far as I can see, the symmetric case is faster than the general case because it only needs to read a half of the shape data. However, I do not know if it is worth being implemented for GPU, since the computing task is not reduced. As for the even-odd method, the multiplying operations are halved, but there are more adding operations. I do not know how to choose the most efficiency method in different cases.

I would think that the symmetric case would make sense on the GPU as well, because the data access is often a limiting factor, but it depends on the loop layout whether the loads are the actual limit. And on a GPU, we have the additional limitation that the mapping to threads within a warp needs to be done by the compiler, and the banking of the memory access, so I think it would require some work to identify a version that is uniformly better. This is simpler with the current CPU code.

Regarding even-odd, I would expect additional problems, because we would need additional synchronization points if we parallelize within the one-dimensional interpolation, again adding complications in the local data access. So here I see the same issue, one would need to carefully lay out the kernels and do associated benchmarking.

To sum up, I would say that the current implementation, using the general case, is fine, and we should focus on getting that to the expected performance limits, before trying to look into the more specialized methods.

**Comment by tjhei** [comment]
Thank you for looking into this!

We certainly will need to work on test coverage.

Regarding the storage type of the local array: My understanding is that local arrays are unlikely to be optimized by the compiler to be in registers, so they will use local memory (which is just segmented shared memory). So, if you can reuse entries between threads, put it into shared memory, otherwise leave it to the compiler.

**Comment by tjhei** [comment]
Sorry, after rereading I see it would be put into global memory.

**Comment by Rombur** [comment]
@YiminJin if you care about performance on the GPU, you need to run your code with [apex](https://github.com/UO-OACISS/apex/wiki/Using-APEX-with-Kokkos#runtime-auto-tuning-support) to auto-tune the code. There are some Kokkos parameters that need to be tuned. If you don't do it, the occupancy on AMD GPUs  will be less than 20% unless you use very high order.

**Comment by YiminJin** [comment]
@kronbichler Thank you very much for answering my questions! 

> Are you saying that you measured the old code and your suggested change, and did not observe a difference in performance? I guess the important question to check is whether the performance reached some point of saturation before/after your change.

Yes. I timed the solve() function in step-64 and did not observe a difference in performance before and after the modification.  I haven't measured the bandwidth yet, but I guess the percentage of bandwidth utilization is similar to the master branch.

> And on a GPU, we have the additional limitation that the mapping to threads within a warp needs to be done by the compiler, and the banking of the memory access, so I think it would require some work to identify a version that is uniformly better. 

I am concerned about the warp issues too. I do not know if Kokkos has taken steps to avoid bank conflicts internally.

> Regarding even-odd, I would expect additional problems, because we would need additional synchronization points if we parallelize within the one-dimensional interpolation, again adding complications in the local data access.

I haven't thought about the effects of synchronization on the performance. I thought the team_barrier was something like MPI_Barrier so I put one after each parallel_for for safety. So, if I remove the indispensable synchronizations and avoid in-place operations by using more temporary arrays, would the performance be improved? I think it is worth trying.

> we should focus on getting that to the expected performance limits, before trying to look into the more specialized methods

I agree with the conclusion. I shall find some performance monitoring tools and make more tests.

**Comment by YiminJin** [comment]
@tjhei Thank you for answering my question! I shall try to make more use of the scratch pad in future tests.

Now I have tested most of the modifications with dim = 2 and 3. The only untested codes are the evaluation and integration for standard spectral element methods ( `Portable::FEEvaluationImplCollocation` in include/deal.II/portable_evaluation_kernels.h).

**Comment by YiminJin** [comment]
@Rombur Thank you very much for introducing apex! It is exactly what I need at present.

**Comment by masterleinad** [comment]
> Regarding the storage type of the local array: My understanding is that local arrays are unlikely to be optimized by the compiler to be in registers, so they will use local memory (which is just segmented shared memory). So, if you can reuse entries between threads, put it into shared memory, otherwise leave it to the compiler.

I would assume that local arrays show in the size of the functor and have a better chance to be optimized depending on the register pressure. Thus, by defualt I would avoid using shared memory if not necessary for sharing between threads unless the limiting factor is spilling.

**Comment by YiminJin** [comment]
@masterleinad Thank you for the advice. In my modifications there are two different situations for the usage of temporary array:
1. Evaluation/Integration without collocation method. In this case we need one or two temporary arrays to reuse the intermediate results, and the temporary arrays act as _input_ values in the tensor contract, which means that each entry is visited by `n_q_points_1d` threads. This case is exactly the same as the example in Kokkos lectures for scratch memory, so I think shared memory is the right choice.
2. Evaluation/Integration with collocation method. In this case I follow the original codes to use in-place implementations, i.e. to use the same array to store the input and output values in the tensor contract (this reduces the temporary storage in 3d). Here the temporary array stores the _output_ values, which means that each entry is visited by only one thread in the tensor contract and, as you suggested, may not be suitable to be put into shared memory.

I would like to do some tests before modifying the codes, yet I am still struggling to find a way to measure the computing efficiency. Timing the `solve()` function in step-64 is too rough for this purpose...

**Comment by masterleinad** [comment]
> I would like to do some tests before modifying the codes, yet I am still struggling to find a way to measure the computing efficiency. Timing the `solve()` function in step-64 is too rough for this purpose...

`Kokkos Tools` (https://github.com/kokkos/kokkos-tools) provides a bunch of tools to profile calls using `Kokkos`. `Space Time Stack` (https://github.com/kokkos/kokkos-tools/wiki/Space-Time-Stack) gives you a good overview over all the time spent in different kernels. Since we are pretty much only using one kernel in the portable matrix-free implementation, you might need to look into using vendor tools like Nsight Compute to understand what's going on inside a kernel.

**Comment by YiminJin** [comment]
@masterleinad Thank you very much! With your suggestions I noticed that the performance evaluation tools are introduced in Kokkos lecture 7, while I have only learnt lectures 1 to 4... No shortcuts to learning indeed.

**Comment by masterleinad** [comment]
> @masterleinad Thank you very much! With your suggestions I noticed that the performance evaluation tools are introduced in Kokkos lecture 7, while I have only learnt lectures 1 to 4... No shortcuts to learning indeed.

There really isn't any reason not to look at that lecture earlier; it doesn't rely on anything else.

**Comment by tjhei** [comment]
@YiminJin What is your plan regarding this PR? It's good to do performance measurements but we should probably also get this PR merged regardless.

Would you mind fixing the error:
> /jenkins/workspace/dealii-mpi_PR-17957/include/deal.II/matrix_free/portable_matrix_free.templates.h:678:5: error: ‘unsigned int dealii::Portable::{anonymous}::compute_scratch_pad_size(dealii::internal::MatrixFreeFunctions::ElementType, int, int, int)’ defined but not used [-Werror=unused-function]

**Comment by YiminJin** [comment]
@tjhei I have to admit that I did not carry out formal tests for this PR. I had hardly begun the testing when I found that the GPU occupancy is really poor in cycles 0 and 1 of step-64 (because the number of threads required for FE evaluation is much smaller than the total number of threads available), which is a serious issue for my research plan (dealing with geometric multigrid). So I have spent the recent three weeks trying to figure out this problem.

I am glad to know that this PR is qualified for being merged. I have thought that there were a huge amount of things to do to make the portable MF functions really applicable, e.g. functions for face assembly, for special FE spaces, for triangular elements, etc. Thus, it would be a little dangerous to build all these stuffs on the structure designed by a beginner of Kokkos. But of course, I would be thrilled if the codes could be merged, and I will continue to work on it.

I am confused about the error: the function `compute_scratch_pad_size()` is called by `MatrixFree::internal_reinit()` in line 765 of include/deal.II/matrix_free/portable_matrix_free.template.h, and the error never occurred on my machine. Is it related with the anonymous namespace?

**Comment by tjhei** [comment]
> So I have spent the recent three weeks trying to figure out this problem.

This is with a very small number of cells, I assume? Of course you won't get good efficiency there. It is just not enough work to begin with to parallelize. This should not be a problem though:
1. Only a very small fraction of the runtime will be spend on coarse levels of the GMG hierarchy (remember, in 3d the amount of work gets cut by a factor of 8 on each level!).
2. One would probably choose a coarse level that is still quite fine and still see good performance. Afterwards, you probably can solve directly or call into AMG.

**Comment by tjhei** [comment]
> I have thought that there were a huge amount of things to do to make the portable MF functions really applicable

Yes. One step at a time. :-)

The original parallel, adaptive, matrix-free framework took 10+ years to be built.

**Comment by tjhei** [comment]
> I am confused about the error: the function `compute_scratch_pad_size()` is called by `MatrixFree::internal_reinit()` in line 765 of include/deal.II/matrix_free/portable_matrix_free.template.h, and the error never occurred on my machine. Is it related with the anonymous namespace?

What happens if you add ``inline`` to the function declaration?

**Comment by YiminJin** [comment]
>This is with a very small number of cells, I assume? Of course you won't get good efficiency there. It is just not enough work to begin with to parallelize. This should not be a problem though

Yes. I agree with the conclusion that this would not be a problem in many cases, especially when the coarsest level is not so coarse. However, what concerns me here is: 1. the load balance problem for adaptively refined mesh, and 2.  the multiple synchronizations in the multigrid algorithm. I have learnt from your paper (2020) that issue 1 is not a big problem in practice, but I still hope to see if there is some way to address both issues (at a tolerable cost of convergence rate) and give full play to the advantage of GPU.

**Comment by YiminJin** [comment]
> What happens if you add inline to the function declaration?

I think the error can definitely be solved if the function is inlined, but it seems weird to declare an inline function which is abstracted because it is complex enough (it would be complex when more situations are taken into account). So I manually "inline" the function by inserting it into the function who calls it, as it is quite simple at present. I cannot test the modification on my machine (because the error never occurred), so I have to submit it and see what the checkers say.

**Comment by tjhei** [comment]
> but it seems weird to declare an inline function which is abstracted because it is complex enoug

The ``inline`` keyword has little to do with actually inlining the code (the compiler decides this independently), but "the meaning of the keyword inline for functions came to mean "multiple definitions are permitted" rather than "inlining is preferred" since C++98", see https://en.cppreference.com/w/cpp/language/inline
So, keeping it as an inline function is the right thing to do.

**Comment by tjhei** [comment]
can you please rebase and fix the compile errors?

**Comment by YiminJin** [comment]
> The inline keyword has little to do with actually inlining the code (the compiler decides this independently), but "the meaning of the keyword inline for functions came to mean "multiple definitions are permitted" rather than "inlining is preferred" since C++98"

Thanks for the instruction. I did not know that. I have restored the helper function `compute_scratch_pad_size` and made it inline.

**Comment by tjhei** [comment]
Can you please squash and rebase to the current master (we do not like having merge commits in PRs)?

**Comment by YiminJin** [comment]
> Can you please squash and rebase to the current master (we do not like having merge commits in PRs)?

Sorry I made some mistakes when trying to rebase yesterday. I tried again this morning,  but this PR is over 200 comments behind the current master and is too hard to squash and rebase. So I  submitted a new PR (#18124) with the same modifications. I shall close this PR if the new one goes well.

**Comment by masterleinad** [comment]
Superseded by #18124.
